### PR TITLE
Issue 2542 : OffHeap Store Iterators

### PR DESCRIPTION
**Pull request by chrisdennis**

### CODE REVIEW

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java [hunk 1]
```diff
@@ -79,7 +79,11 @@ public boolean equals(Object o, ByteBuffer byteBuffer) {
     long lastAccessTime = byteBuffer.getLong();
     long expireTime = byteBuffer.getLong();
     byteBuffer.getLong(); // hits read from disk. It is kept for compatibility reasons with previously saved data
-    return new LazyOffHeapValueHolder<>(id, byteBuffer.slice(), serializer,
+
+    ByteBuffer attached = byteBuffer.slice();
+    ByteBuffer detached = ByteBuffer.allocateDirect(attached.remaining());
```
Comment (lorban) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java]: Are you sure about the direct byte buffer? A heap one would put less strain on the GC as it has no finalizer, and I don't see the benefit of a offheap one here.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java [hunk 2]
```diff
@@ -79,7 +79,11 @@ public boolean equals(Object o, ByteBuffer byteBuffer) {
...
+    detached.put(attached).clear();
```
[the remark this answers is not in the document]
Comment (lorban) [reply impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java]: Also, this extra memory copy did not previously exist.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java (hunk 1 above)
Comment (chrisdennis) [reply impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java]: A heap one is probably a better choice you're right.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisdennis) [reply impl/src/main/java/org/ehcache/impl/internal/store/offheap/portability/OffHeapValueHolderPortability.java]: Right - it's the lack of a detaching copy that is causing issues in the iterators. Whats happening at the moment is when we read the value from offheap we literally just grab a reference... then we either deserialize, detach (and then potentially deserialize later) or drop the value before releasing the lock (depending on what we're doing).  The danger is that if we don't detach before releasing the lock and then subsequently deserialize we can read invalid state. This path illustrates the simplest and safest fix (which may come with a measurable performance dip) which is to always do a detaching copy.

**Pull request closed without merging**